ast/functions.go

```diff
@@ -382,3 +402,219 @@ func (n *AggregateFuncExpr) Accept(v Visitor) (Node, bool) {
 	}
 	return v.Leave(n)
 }
+
+// Update is used for update aggregate context.
+func (n *AggregateFuncExpr) Update() error {
+	name := strings.ToLower(n.F)
+	switch name {
+	case AggFuncCount:
+		return n.updateCount()
+	case AggFuncFirstRow:
+		return n.updateFirstRow()
+	}
+	return nil
+}
+
+// GetContext gets aggregate evaluation context for the current group.
+// If it is nil, add a new context into ContextPerGroup.
+func (n *AggregateFuncExpr) GetContext() *AggEvaluateContext {
+	if n.ContextPerGroup == nil {
+		n.ContextPerGroup = make(map[string](*AggEvaluateContext))
+	}
+	if _, ok := n.ContextPerGroup[n.CurrentGroup]; !ok {
+		c := &AggEvaluateContext{}
+		if n.Distinct {
+			c.Distinct = CreateAggregateDistinct(n.F, n.Distinct)
+		}
+		n.ContextPerGroup[n.CurrentGroup] = c
+	}
+	return n.ContextPerGroup[n.CurrentGroup]
+}
+
+func (n *AggregateFuncExpr) updateCount() error {
+	ctx := n.GetContext()
+	var value interface{}
+	if len(n.Args) > 0 {
+		hasNil := false
+		for _, a := range n.Args {
+			value = a.GetValue()
+			if value == nil {
+				hasNil = true
+			}
+		}
+		if hasNil {
+			return nil
+		}
+	}
+	if n.Distinct {
+		// TODO: compose values into value. For example select count(DISTINCT c1, c2) from t;
+		d, err := ctx.Distinct.IsDistinct(value)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if !d {
+			return nil
+		}
+	}
+	ctx.Count++
+	return nil
+}
+
+func (n *AggregateFuncExpr) updateFirstRow() error {
+	ctx := n.GetContext()
+	if ctx.evaluated {
+		return nil
+	}
+	if len(n.Args) != 1 {
+		return errors.New("Wrong number of args for AggFuncFirstRow")
+	}
+	ctx.Value = n.Args[0].GetValue()
+	ctx.evaluated = true
+	return nil
+}
+
+// AggFuncDetector visits Expr tree to check if it contains AggregateFuncExpr.
+type AggFuncDetector struct {
+	HasAggFunc bool
+	detecting  bool
+}
+
+// Enter implements Visitor interface.
+func (a *AggFuncDetector) Enter(n Node) (node Node, skipChildren bool) {
+	defer func() {
```

Comment:
defer is too expensive here in `Enter` because it will be called many times during visiting.